Tests/MongoSwiftSyncTests/UnifiedTestRunner/Matching.swift

```diff
@@ -0,0 +1,328 @@
+import MongoSwiftSync
+import TestsCommon
+
+extension UnifiedOperationResult {
+    /// Determines whether this result matches `expected`.
+    func matches(expected: BSON, entities: EntityMap) throws -> Bool {
+        let actual: MatchableResult
+        switch self {
+        case let .bson(bson):
+            actual = MatchableResult(from: bson)
+        case let .rootDocument(doc):
+            actual = .rootDocument(doc)
+        case let .rootDocumentArray(arr):
+            actual = .rootDocumentArray(arr)
+        case .none:
+            actual = .none
+        default:
+            return false
+        }
+
+        return try actual.matches(expected, entities: entities)
+    }
+}
+
+/// Enum representing types that can be matched against expected values.
+enum MatchableResult {
+    /// A root document. i.e. a documents where extra keys are ignored when matching against an expected document.
+    case rootDocument(BSONDocument)
+    /// An array of root documents.
+    case rootDocumentArray([BSONDocument])
+    /// A (non-root) document.
+    case subDocument(BSONDocument)
+    /// An array of BSONs.
+    case array([BSON])
+    /// A non-document, non-array BSON.
+    case scalar(BSON)
+    /// A nil result.
+    case none
+
+    /// Initializes an instance of `MatchableResult` from a `BSON`.
+    init(from bson: BSON?) {
+        guard let bson = bson else {
+            self = .none
+            return
+        }
+        switch bson {
+        case let .document(doc):
+            self = .subDocument(doc)
+        case let .array(arr):
+            self = .array(arr)
+        default:
+            self = .scalar(bson)
+        }
+    }
+
+    /// Determines whether `self` matches `expected`, recursing if needed for nested documents and arrays.
+    fileprivate func matches(_ expected: BSON, entities: EntityMap) throws -> Bool {
```

Comment:
We discussed some options for plumbing down a key path or returning a more detailed result here so that the error output is more useful. I think those things would be really nice to have, but in the interests of moving this project along I think they can wait until another PR. Is that the current plan? I honestly don't remember where we left that discussion.

Reply:
honestly, I don't remember either 🙂  I think I maybe said I would investigate it but then didn't do so. I think tackling it in a follow up PR makes sense